File: webdriver/pageobjects/page_vaults.py
#!python3
from page_objects import PageObject, PageElement,MultiPageElement
import storedsafe_driver_values as constants
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.support.ui import Select
from selenium.webdriver.common.alert import Alert
from pageobjects.element_top_menu import top_menu
from pageobjects.element_bottom_infobar import bottom_infobar

import selenium.selenium

class PageVaults(PageObject):
    driver = None
    topmenu = None
    bottommenu = None
    vault_list = None
    vault_content_list = []


    def __init__(self,webdriver):
        self.driver = webdriver
        PageObject.__init__(self,self.driver)
        #self.button_logout = self.driver.find_element_by_css_selector('input[value="Logout"]')
        self.topmenu = top_menu(self.driver)
        self.bottommenu = bottom_infobar(self.driver)
        self.vault_list = self.driver.find_elements(By.XPATH, '//*[starts-with(@id,"bartitle")]')
        self.new_vault_button = self.driver.find_element_by_id('newgroup')
        for i in self.vault_list:
            p = "".join(list(filter(str.isdigit,i.get_attribute("id"))))
            z =self.driver.find_element(By.ID, 'bar'+p)
            self.vault_content_list.append(z)

    def verify_on_vaults_page(self):
        try:
            self.driver.find_element_by_id('objectlistwindow')
            v_ex = True
        except:
            v_ex = False

        return str(self.driver.current_url).startswith(constants.url_base+constants.url_vault) and v_ex


    def logout(self):
        import pageobjects.page_login as pageLogin
        self.topmenu.logout()
        return pageLogin.PageLogin(self.driver)

    def close_vault(self,vaultname):
        id = constants.get_vaultsid(vaultname)
        for v in self.vault_list:
            if v.get_attribute('id') == "bartitle{}".format(id):
                if v.get_attribute('class') == 'bars _on':
                    v.click()
                    return

    def open_vault(self,vaultname):
        try:
            id = constants.get_vaultsid(vaultname)
        except KeyError:
            id=constants.db_handler.get_new_vault_id(vaultname)

        return self.open_vault_by_id(id)

    def open_new_vault(self,vaultname):
        id = constants.db_handler.get_new_vault_id(vaultname)
        return self.open_vault_by_id(id)

    def open_vault_by_id(self,vaultid):
        for v in self.vault_list:
            if v.get_attribute('id') == 'bartitle' + str(vaultid):
                if v.get_attribute('class') == 'bars _off':
                    v.click()
                return True
        return False

    def create_new_item_server(self,vaultname,itemname,iteminfo):
        #split in multiple lesser classes to better handle multiple object types
        try:
            self.open_vault(vaultname)
            v_id = constants.get_vaultsid(vaultname)
            self.driver.find_element_by_id('bar'+v_id+'add').click()
            self.driver.find_element_by_id('templateid1').click()
            self.driver.find_element_by_id('cont'+v_id).click()
            self.driver.find_element_by_id('host').send_keys(itemname)
            self.driver.find_element_by_id('username').send_keys(iteminfo[1])
            if iteminfo[2] == '':
                self.driver.find_element_by_id('gen').click()
            else:
                self.driver.find_element_by_id('password').send_keys(iteminfo[2])
            if iteminfo[3] == 'True':
                self.driver.find_element_by_id('password_alarm').click()
            self.driver.find_element_by_id('info').send_keys(iteminfo[4])
            self.driver.find_element_by_id('cryptedinfo').send_keys(iteminfo[5])
            self.driver.find_element_by_id('submitbutton').click()
        except:
            return False
        return True

    def mark_object(self,vaultname, objectname):
        try:
            self.open_vault(vaultname)
            vaultid = constants.get_vaultsid(vaultname)
            try:
                object= constants.get_object(objectname)
                o_id = object[0]
                o_type = object[2]
            except KeyError:
                o_id = constants.db_handler.get_new_created_item_in_vault_by_name(objectname,vaultid)
                o_type = constants.db_handler.get_new_object_type(o_id,vaultid)


            self.driver.find_element_by_id('mod_{}_{}_{}'.format(vaultid,o_type,o_id)).click()
        except:
            import sys
            print(sys.exc_info())
            return False
        return True

    def copy_object(self,vaultname,objectname):
        a = self.mark_object(vaultname,objectname)
        if not a:
            return False
        try:
            self.driver.find_element_by_id('copy_'+constants.get_vaultsid(vaultname)).click()
        except:
            return False
        return True

    def move_object(self,vaultname,objectname):
        a = self.mark_object(vaultname, objectname)
        if not a:
            return False
        try:
            self.driver.find_element_by_id('move_' + constants.get_vaultsid(vaultname)).click()
        except:
            return False
        return True

    def get_encrypted_data(self,vaultid,objectid):
        self.open_vault_by_id(vaultid)
        obj=self.driver.find_element(By.XPATH, '//*[contains(@id,":{}:")]'.format(objectid))
        obj.click()
        data = obj.find_element_by_class_name("obfuscate").text
        return data

    def paste_object(self,vaultname):
        try:
            self.open_vault(vaultname)
            try:
                vaultid = constants.get_vaultsid(vaultname)
            except KeyError:
                vaultid = constants.db_handler.get_new_vault_id(vaultname)
            self.driver.find_element_by_id('paste_{}'.format(vaultid)).click()
            Alert(self.driver).accept()
        except:
            return False
        #wait for waitwondow to disappear => wait for database to update
        WebDriverWait(self.driver, 10).until(EC.invisibility_of_element_located((By.ID, "waitwindow")))

        return True


    def create_new_vault(self,vaultname):
        try:
            vaultinfo = constants.get_newvault_info(vaultname)
            self.new_vault_button.click()
            self.driver.find_element_by_id('groupname').send_keys(vaultname)
            Select(self.driver.find_element_by_id('policy')).select_by_visible_text(vaultinfo[0])
            self.driver.find_element_by_id('info').send_keys(vaultinfo[1])
            self.driver.find_element_by_id('submitbutton').click()
        except:
            return False
        return True

    def delete_vault(self,vaultid):
        try:
            self.open_vault_by_id(vaultid)
            self.driver.find_element_by_id("bar{}edit".format(vaultid)).click()
            WebDriverWait(self.driver, 10).until(EC.visibility_of_element_located((By.ID, "popupwindow")))
            self.driver.find_element_by_id("deletebutton").click()

            Alert(self.driver).accept()
            WebDriverWait(self.driver, 10).until(EC.invisibility_of_element_located((By.ID, "popupwindow")))
        except:
            return False
        return True


    def delete_object(self,vaultid,objectid,objecttype):
        try:
            self.open_vault_by_id(vaultid)
            self.driver.find_element_by_id("link-{}".format(objectid)).click()
            if(objecttype!='8' or objecttype != '9'):
                self.driver.find_element_by_id("editbtn#{}".format(objectid)).click()
            WebDriverWait(self.driver, 10).until(EC.visibility_of_element_located((By.ID, "popupwindow")))
            self.driver.find_element_by_id('deletebutton').click()
            Alert(self.driver).accept()
            WebDriverWait(self.driver, 10).until(EC.invisibility_of_element_located((By.ID, "popupwindow")))
        except:
            return False
        return True

    def leave_vault(self,userid,vaultid):
        try:
            self.open_vault_by_id(vaultid)
            self.driver.find_element_by_id("bar{}users2".format(vaultid)).click()
            WebDriverWait(self.driver,10).until(EC.visibility_of_element_located((By.ID,"btnid{}".format(userid))))
            self.driver.find_element_by_id("btnid{}".format(userid)).click()
            Alert(self.driver).accept()
            WebDriverWait(self.driver,10).until((EC.invisibility_of_element_located((By.ID, "waitwindow"))))
        except:
            return False
        return True

File: capybara/features/support/page_objects/page_vaults.rb
class Page_vaults < Page_logged_in
  def self.isAtVaultsPage
    #verify on right path
    if Capybara.app_host+'groups.php' != current_url.split('#')[0]
      return false
    end
    #verify on a page with a div where id=objectlistwindow
    return has_css?("[id='objectlistwindow']")
  end
  def self.isVaultInList(v_id)
    begin
      has_css?("[id = 'bartitle#{v_id}']")
    rescue
      return false
    end
    return true
  end

  def self.openVault(v_id)
    if !isVaultOpen?(v_id)
      find(:css, "[id = 'bartitle#{v_id}']").click
    end
  end

  def self.closeVault(v_id)
    if isVaultOpen?(v_id)
      find(:css, "[id = 'bartitle#{v_id}']").click
    end
  end

  def self.isVaultOpen?(v_id)
    return find(:css, "[id = 'bartitle#{v_id}']")['class'] == 'bars _on'
  end

  def self.isLoggedInAsAdmin
    has_css?("[id = 'newgroup']") && has_css?("[id = 'users']")
  end
  def self.isLoggedInAsWrite
    has_css?("[id = 'newgroup']") && !has_css?("[id = 'users']")
  end
  def self.isLoggedInAsRead
    !has_css?("[id = 'newgroup']") && !has_css?("[id = 'users']")
  end
  def self.hasAudit
    return has_css?("[id='audit']")
  end
  def self.hasUgList
    #how?
    #divide on admin, write, read?
    return true
  end

  def self.isObjectInVault(vaultid,objectname)
    return find(:css,"[id=bar#{vaultid}]").has_content?(objectname)
  end

  def self.isObjectNotInVault(vaultid,objectname)
    return find(:css,"[id=bar#{vaultid}]").has_no_content?(objectname)
  end

  def self.create_new_item_server(vaultid, itemname)
    openVault(vaultid)
    click_button "bar#{vaultid}add"
    choose 'templateid1'
    click_button "cont#{vaultid}"
    itemdata = C_Support.get_newitem_server(itemname)
    fill_in 'host',:with=> itemname
    fill_in 'username',:with=> itemdata[1]
    if itemdata[2]== ''
      click_button 'gen'
    else
      fill_in 'password',:with => itemdata[2]
    end
    if itemdata[3]
      check 'password_alarm'
    end
    fill_in 'info',:with=>itemdata[4]
    fill_in 'cryptedinfo',:with=>itemdata[5]
    click_button 'submitbutton'
    return true
  end

  def self.create_new_vault(vaultname)
    vaultdata = C_Support.get_newvault_info(vaultname)
    click_button "newgroup"
    fill_in 'groupname',:with=>vaultname
    select vaultdata[0], :from=> 'policy'
    fill_in 'info',:with=>vaultdata[1]
    click_button 'submitbutton'
    return true
  end

  def self.read_encrypted_data(vaultname,objectname)
    v_id = C_Support.get_vault_id(vaultname)
    self.openVault(v_id)
    o_id = C_Support.get_object_id(v_id,objectname)
    decrypted = false
    if has_css? ("[id*=':#{o_id}:']")
      el = find(:css,"[id*=':#{o_id}:']")
      el .click
      decrypted = el.find(:css,"[class='obfuscate']").text
    end
    return decrypted
  end

  def self.copy_object(vault_from_id,objectname)
    self.openVault(vault_from_id)
    o_type = C_Support.get_object_type(objectname)
    objectid = C_Support.get_object_id(vault_from_id,objectname)
    check "mod_#{vault_from_id}_#{o_type}_#{objectid}"
    find(:css, "[id='copy_#{vault_from_id}']") .click
    return true
  end

  def self.cut_object(vault_from_id,objectname)
    self.openVault(vault_from_id)
    o_type = C_Support.get_object_type(objectname)
    objectid = C_Support.get_object_id(vault_from_id,objectname)
    check "mod_#{vault_from_id}_#{o_type}_#{objectid}"
    find(:css, "[id='move_#{vault_from_id}']") .click
    return true
  end

  def self.paste_object(vault_from_id,vault_to_id)
    self.openVault(vault_to_id)
    find(:css, "[id='paste_#{vault_to_id}']") .click
    page.driver.browser.switch_to.alert .accept

    self.closeVault(vault_from_id)
    self.closeVault(vault_to_id)
    self.openVault(vault_to_id)

    return true
  end

  def self.delete_object(vaultid,objectid,objecttype)
    self.openVault(vaultid)
    find(:css, "[id='link-#{objectid}']") .click
    if(objecttype != '8' and objecttype != '9')
      #if enters workbench
      click_button "editbtn##{objectid}"
      #find("editbtn##{objectid}") .click
    end
    click_button "deletebutton"
    page.driver.browser.switch_to.alert .accept

    self.isAtVaultsPage
    self.closeVault(vaultid)
    self.openVault(vaultid)
    return true
  end

  def self.delete_vault(vaultid)
    #assumes user is admin member of vault!
    self.openVault(vaultid)
    click_button "bar#{vaultid}edit"
    click_button "deletebutton"
    page.driver.browser.switch_to.alert .accept

    return true
  end

  def self.verify_delete_vault_failed(vaultid)

    val1 = self.isAtVaultsPage()
    val2 = find_by_id('errorwindow').text == "Vault cannot be deleted with active items"
    val3 = self.isVaultInList(vaultid)

    return val1 && val2 && val3
  end

end
